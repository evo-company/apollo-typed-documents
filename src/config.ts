export type TypedFilesModulesConfig = {
  /**
   * @description Import location for types generated by typescript operations
   * plugin.
   */
  typesModule: string;

  /**
   * @default false
   * @description By default, only the filename is used to generate TS module
   * declarations. Setting this to `true` will generate it with a full path
   * based on the CWD.
   */
  relativeToCwd?: boolean;

  /**
   * @default "*\/"
   * @description By default all module paths will be prefixed with "*\/".
   * Override this to remove or specify a custom prefix (helpful to set to "" in
   * conjunction with relativeToCwd.)
   */
  prefix?: string;

  /**
   * @default ""
   * @description Used in conjunction with relativeToCwd if if imports should be
   * relative to some subdirectory within the cwd, e.g. "src".
   */
  stripPrefix?: string;

  /**
   * @default "@graphql-typed-document-node/core"
   * @description Import location for TypedDocumentNode.
   */
  typedDocumentNodeModule?: string;

  /**
   * @default false
   * @description Exclude default exports from modules.
   */
  excludeDefaultExports?: boolean;

  /**
   * @default undefined
   * @description Add suffix to generated operation result type names
   * e.g we have operation name Author and suffix QueryType, we will get AuthorQueryType
   */
  operationResultSuffix?: string | null;

  /**
   * @default false
   * @description use operation name as suffix
   * e.g we have operation query Author, we will get AuthorQuery
   * e.g we have operation mutation Author? we will get AuthorMutation
   */
  useOperationNameAsSuffix?: boolean;
};
