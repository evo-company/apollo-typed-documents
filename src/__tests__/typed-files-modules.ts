import { codegen } from "@graphql-codegen/core";
import { Types } from "@graphql-codegen/plugin-helpers";
import { buildSchema, parse, printSchema } from "graphql";

import { TypedFilesModulesConfig } from "../config";
import * as typedFilesModules from "../index";

const schema = buildSchema(`
  type Author {
    idField: ID!
  }

  type Query {
    authors: [Author]
  }

  type Mutation {
    createAuthor: Author!
  }

  schema {
    query: Query
    mutation: Mutation
  }
`);

const getConfig = (
  generateOptions: Partial<Types.GenerateOptions> = {},
  pluginOptions: Partial<TypedFilesModulesConfig> = {}
): Types.GenerateOptions => ({
  filename: "ops.d.ts",
  schema: parse(printSchema(schema)),
  plugins: [
    {
      typedFilesModules: {
        typesModule: "@codegen-types",
        typedDocumentNodeModule: "docnode",
        ...pluginOptions,
      },
    },
  ],
  pluginMap: { typedFilesModules },
  config: {},
  documents: [],
  ...generateOptions,
});

const singleOpDocuments = [
  {
    location: "authors.gql",
    document: parse(`
      query authors {
        authors {
          idField
        }
      }
    `),
  },
  {
    location: "createAuthor.gql",
    document: parse(`
      mutation createAuthor {
        createAuthor {
          idField
        }
      }
    `),
  },
];

const multiOpDocuments = [
  {
    location: "authors.gql",
    document: parse(`
      query authors {
        authors {
          idField
        }
      }
      query alsoAuthors {
        authors {
          idField
        }
      }
    `),
  },
  {
    location: "createAuthor.gql",
    document: parse(`
      mutation createAuthor {
        createAuthor {
          idField
        }
      }
      mutation alsoCreateAuthor {
        createAuthor {
          idField
        }
      }
    `),
  },
];

describe("typed-files-modules", () => {
  it("should not have any output when there are no documents", async () => {
    const config = getConfig();
    const output = await codegen(config);

    expect(output).toMatchInlineSnapshot(`""`);
  });

  it("should have ambient module declarations for each document", async () => {
    const config = getConfig({ documents: singleOpDocuments });
    const output = await codegen(config);

    expect(output).toMatchSnapshot();
  });

  it("should not have default exports for multiple operations", async () => {
    const config = getConfig({ documents: multiOpDocuments });
    const output = await codegen(config);

    expect(output).toMatchSnapshot();
  });

  it("should not have default exports if specifically excluded", async () => {
    const config = getConfig(
      { documents: singleOpDocuments },
      { excludeDefaultExports: true }
    );
    const output = await codegen(config);
    expect(output).toMatchSnapshot();
  });

  describe("module path customization", () => {
    const documents = singleOpDocuments;

    it("wildcards the basename by default", async () => {
      const config = getConfig({ documents });
      const output = await codegen(config);

      expect(output).toMatchSnapshot();
    });

    it("respects the relativeToCwd setting", async () => {
      const config = getConfig({ documents }, { relativeToCwd: true });
      const output = await codegen(config);

      expect(output).toEqual(
        expect.stringContaining(`declare module "*/authors.gql"`)
      );
      expect(output).toEqual(
        expect.stringContaining(`declare module "*/createAuthor.gql"`)
      );
    });

    it("respects the prefix setting", async () => {
      const config = getConfig({ documents }, { prefix: "gql/" });
      const output = await codegen(config);

      expect(output).toEqual(
        expect.stringContaining(`declare module "gql/authors.gql"`)
      );
      expect(output).toEqual(
        expect.stringContaining(`declare module "gql/createAuthor.gql"`)
      );
    });

    it("allows combining path settings", async () => {
      const config = getConfig(
        { documents },
        { prefix: "defs/", relativeToCwd: true }
      );

      const output = await codegen(config);

      expect(output).toEqual(
        expect.stringContaining(`declare module "defs/authors.gql"`)
      );
      expect(output).toEqual(
        expect.stringContaining(`declare module "defs/createAuthor.gql"`)
      );
    });
  });

  describe("typedDocumentNodeModule", () => {
    it("defaults to graphql code generator's definition", async () => {
      const config = getConfig(
        { documents: singleOpDocuments },
        { typedDocumentNodeModule: undefined }
      );

      const output = await codegen(config);
      expect(output).toEqual(
        expect.stringContaining(
          `import { TypedDocumentNode } from "@graphql-typed-document-node/core";`
        )
      );
    });

    it("can be customized", async () => {
      const config = getConfig(
        { documents: singleOpDocuments },
        { typedDocumentNodeModule: "@apollo/client" }
      );
      const output = await codegen(config);
      expect(output).toEqual(
        expect.stringContaining(
          `import { TypedDocumentNode } from "@apollo/client";`
        )
      );
    });
  });
});
