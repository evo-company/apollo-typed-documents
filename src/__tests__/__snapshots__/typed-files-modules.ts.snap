// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typed-files-modules module path customization wildcards the basename by default 1`] = `
"declare module \\"*/authors.gql\\" {
  import { TypedDocumentNode } from \\"docnode\\";
  import { AuthorsQuery, AuthorsQueryVariables } from \\"@codegen-types\\";
  export const authors: TypedDocumentNode<AuthorsQuery, AuthorsQueryVariables>;
  export default authors;
}

declare module \\"*/createAuthor.gql\\" {
  import { TypedDocumentNode } from \\"docnode\\";
  import { CreateAuthorMutation, CreateAuthorMutationVariables } from \\"@codegen-types\\";
  export const createAuthor: TypedDocumentNode<CreateAuthorMutation, CreateAuthorMutationVariables>;
  export default createAuthor;
}"
`;

exports[`typed-files-modules should have ambient module declarations for each document 1`] = `
"declare module \\"*/authors.gql\\" {
  import { TypedDocumentNode } from \\"docnode\\";
  import { AuthorsQuery, AuthorsQueryVariables } from \\"@codegen-types\\";
  export const authors: TypedDocumentNode<AuthorsQuery, AuthorsQueryVariables>;
  export default authors;
}

declare module \\"*/createAuthor.gql\\" {
  import { TypedDocumentNode } from \\"docnode\\";
  import { CreateAuthorMutation, CreateAuthorMutationVariables } from \\"@codegen-types\\";
  export const createAuthor: TypedDocumentNode<CreateAuthorMutation, CreateAuthorMutationVariables>;
  export default createAuthor;
}"
`;

exports[`typed-files-modules should not have default exports for multiple operations 1`] = `
"declare module \\"*/authors.gql\\" {
  import { TypedDocumentNode } from \\"docnode\\";
  import { AuthorsQuery, AuthorsQueryVariables } from \\"@codegen-types\\";
  export const authors: TypedDocumentNode<AuthorsQuery, AuthorsQueryVariables>;
  import { AlsoAuthorsQuery, AlsoAuthorsQueryVariables } from \\"@codegen-types\\";
  export const alsoAuthors: TypedDocumentNode<AlsoAuthorsQuery, AlsoAuthorsQueryVariables>;
}

declare module \\"*/createAuthor.gql\\" {
  import { TypedDocumentNode } from \\"docnode\\";
  import { CreateAuthorMutation, CreateAuthorMutationVariables } from \\"@codegen-types\\";
  export const createAuthor: TypedDocumentNode<CreateAuthorMutation, CreateAuthorMutationVariables>;
  import { AlsoCreateAuthorMutation, AlsoCreateAuthorMutationVariables } from \\"@codegen-types\\";
  export const alsoCreateAuthor: TypedDocumentNode<AlsoCreateAuthorMutation, AlsoCreateAuthorMutationVariables>;
}"
`;

exports[`typed-files-modules should not have default exports if specifically excluded 1`] = `
"declare module \\"*/authors.gql\\" {
  import { TypedDocumentNode } from \\"docnode\\";
  import { AuthorsQuery, AuthorsQueryVariables } from \\"@codegen-types\\";
  export const authors: TypedDocumentNode<AuthorsQuery, AuthorsQueryVariables>;
}

declare module \\"*/createAuthor.gql\\" {
  import { TypedDocumentNode } from \\"docnode\\";
  import { CreateAuthorMutation, CreateAuthorMutationVariables } from \\"@codegen-types\\";
  export const createAuthor: TypedDocumentNode<CreateAuthorMutation, CreateAuthorMutationVariables>;
}"
`;
